README.md
# timeout-order

订单超时，自动关闭的实现，采用循环队列，参考[这篇文章][link_ref_article]。

# 程序演示
下载代码后，直接运行`com.tuyu.TimeoutOrderApplication`类，默认监听8081端口
```

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.1.3.RELEASE)

2019-02-19 16:48:25.012  INFO 12529 --- [           main] com.tuyu.TimeoutOrderApplication         : Starting TimeoutOrderApplication on scutuyu.local with PID 12529 (/Users/tuyu/Desktop/test/timeout-order/target/classes started by tuyu in /Users/tuyu/Desktop/test/timeout-order)
2019-02-19 16:48:25.015  INFO 12529 --- [           main] com.tuyu.TimeoutOrderApplication         : No active profile set, falling back to default profiles: default
2019-02-19 16:48:26.011  INFO 12529 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8081 (http)
2019-02-19 16:48:26.036  INFO 12529 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2019-02-19 16:48:26.036  INFO 12529 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.16]
2019-02-19 16:48:26.041  INFO 12529 --- [           main] o.a.catalina.core.AprLifecycleListener   : The APR based Apache Tomcat Native library which allows optimal performance in production environments was not found on the java.library.path: [/Users/tuyu/Library/Java/Extensions:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java:.]
2019-02-19 16:48:26.123  INFO 12529 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
2019-02-19 16:48:26.123  INFO 12529 --- [           main] o.s.web.context.ContextLoader            : Root WebApplicationContext: initialization completed in 1064 ms
2019-02-19 16:48:26.173  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 0, next point: 1
2019-02-19 16:48:26.310  INFO 12529 --- [           main] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService 'applicationTaskExecutor'
2019-02-19 16:48:26.386  INFO 12529 --- [           main] o.s.b.a.w.s.WelcomePageHandlerMapping    : Adding welcome page: class path resource [static/index.html]
2019-02-19 16:48:26.460  INFO 12529 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8081 (http) with context path ''
2019-02-19 16:48:26.463  INFO 12529 --- [           main] com.tuyu.TimeoutOrderApplication         : Started TimeoutOrderApplication in 1.835 seconds (JVM running for 2.243)
2019-02-19 16:48:27.176  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 1, next point: 2
2019-02-19 16:48:28.175  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 2, next point: 3
2019-02-19 16:48:29.174  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 3, next point: 4
2019-02-19 16:48:30.176  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 4, next point: 5
2019-02-19 16:48:31.174  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 5, next point: 6
2019-02-19 16:48:32.174  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 6, next point: 7
2019-02-19 16:48:33.176  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 7, next point: 8
2019-02-19 16:48:34.178  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 8, next point: 9
2019-02-19 16:48:35.173  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 9, next point: 10
2019-02-19 16:48:36.174  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 10, next point: 11
2019-02-19 16:48:37.178  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 11, next point: 12
```

程序后每秒钟打印当前指针的索引

终端执行
```shell
curl -X POST --header 'Content-Type: application/x-www-form-urlencoded' --header 'Accept: text/xml' -d 'name=香烟1&price=12220&delay=6' 'http://localhost:8081/order'
```
就可以得到如下的返回
```
订单提交成功
```

程序控制台输出如下：
```
2019-02-19 16:50:01.171  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 95, next point: 96
2019-02-19 16:50:02.172  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 96, next point: 97
2019-02-19 16:50:02.989  INFO 12529 --- [nio-8081-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2019-02-19 16:50:02.989  INFO 12529 --- [nio-8081-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2019-02-19 16:50:02.994  INFO 12529 --- [nio-8081-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 5 ms
2019-02-19 16:50:03.027  INFO 12529 --- [nio-8081-exec-1] com.tuyu.common.CircularQueue            : submit an order: df7a434b89824c6b84918c15f683daeb, timeout in 6 seconds
2019-02-19 16:50:03.028  INFO 12529 --- [nio-8081-exec-1] com.tuyu.common.CircularQueue            : idx: 103
2019-02-19 16:50:03.028  INFO 12529 --- [nio-8081-exec-1] com.tuyu.common.CircularQueue            : set: [CircularQueue.TimingTask(orderUuid=df7a434b89824c6b84918c15f683daeb, order=Order(orderUuid=df7a434b89824c6b84918c15f683daeb, name=香烟1, price=12220, orderTime=null, createTime=null, updateTime=null, timeout=false, payed=false, payTime=null), delay=6, cycleNum=0)]
2019-02-19 16:50:03.028  INFO 12529 --- [nio-8081-exec-1] com.tuyu.common.CircularQueue            : cachedMap: {df7a434b89824c6b84918c15f683daeb=103}
2019-02-19 16:50:03.171  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 97, next point: 98
2019-02-19 16:50:04.173  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 98, next point: 99
2019-02-19 16:50:05.173  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 99, next point: 100
2019-02-19 16:50:06.174  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 100, next point: 101
2019-02-19 16:50:07.172  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 101, next point: 102
2019-02-19 16:50:08.172  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 102, next point: 103
2019-02-19 16:50:09.174  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 103, next point: 104
2019-02-19 16:50:09.174  INFO 12529 --- [pool-2-thread-1] com.tuyu.common.CircularQueue            : timeout task uuid: df7a434b89824c6b84918c15f683daeb
2019-02-19 16:50:09.174  INFO 12529 --- [pool-2-thread-1] com.tuyu.common.CircularQueue            : order timeout: Order(orderUuid=df7a434b89824c6b84918c15f683daeb, name=香烟1, price=12220, orderTime=null, createTime=null, updateTime=Tue Feb 19 16:50:09 CST 2019, timeout=true, payed=false, payTime=null)
2019-02-19 16:50:09.174  INFO 12529 --- [pool-2-thread-1] com.tuyu.common.CircularQueue            : set: []
2019-02-19 16:50:09.175  INFO 12529 --- [pool-2-thread-1] com.tuyu.common.CircularQueue            : cachedMap: {}
2019-02-19 16:50:10.170  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 104, next point: 105
2019-02-19 16:50:11.173  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 105, next point: 106
2019-02-19 16:50:12.171  INFO 12529 --- [pool-1-thread-1] com.tuyu.common.CircularQueue            : cur point: 106, next point: 107
```

提交的订单，在6秒钟后自动被关闭了，并从循环队列中移除。

# Changelog
[learn more about the changelog][link_changelog]

[link_ref_article]: https://blog.csdn.net/verifocus/article/details/79135895
[link_changelog]: https://github.com/scutuyu/timeout-order/blob/master/CHANGELOG.md

CHANGELOG.md
#  (2019-02-19)


### Features

* **订单超时，自动关闭:** 模拟客户端提交一个订单，默认10秒之后，如果订单没有支付，就将订单状态设为关闭状态 ([538747e](https://github.com/scutuyu/timeout-order/commit/538747e))





src/main/java/com/tuyu/common/CircularQueue.java
package com.tuyu.common;

import com.tuyu.entity.Order;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * 环形队列
 * <p>
 *     用环形队列来实现订单超时自动关闭
 * </p>
 *
 * @author tuyu
 * @date 2/18/19
 * Talk is cheap, show me the code.
 */
@Slf4j
@Component
public class CircularQueue {

    /**
     * 循环队列的默认容量
     */
    private static final int DEFAULT_SIZE = 3600;
    /**
     * 缓存环形队列中的所有订单的Map的默认容量
     */
    private static final int DEFAULT_CACHE_SIZE = 1024;

    /** 缓存订单在环形队列中的索引，key=订单uuid，value=索引 */
    private Map<String, Integer> cachedMap;

    /** 环形队列，数组实现 */
    private OrderSet[] table;

    /** 指针 */
    private AtomicInteger point = new AtomicInteger(0);


    /**
     * 构造函数中会new一个线程来轮询环形队列
     */
    public CircularQueue() {
        this.table = new OrderSet[DEFAULT_SIZE];
        this.cachedMap = new HashMap<>(DEFAULT_CACHE_SIZE);
        ScheduledExecutorService scheduledExecutorService = new ScheduledThreadPoolExecutor(1,
                Executors.defaultThreadFactory(), new ThreadPoolExecutor.AbortPolicy());
        scheduledExecutorService.scheduleAtFixedRate(
                new RoundRinTimerTask(this), 0L, 1L, TimeUnit.SECONDS);
    }


    /**
     * 提交一个订单到环形队列中，一定时间内没有付款，该订单就会自动关闭
     *
     * @param order 订单
     * @param delay 订单保留期限，单位: 秒
     */
    public void submitOrder(Order order, int delay) {
        int p = this.point.get() + delay;
        int idx = index(p);
        int cycleNum = cycleNumber(p);
        OrderSet set = null;
        if ((set = table[idx]) == null) {
            set = table[idx] = new OrderSet();
        }
        synchronized (this){
            log.info("submit an order: {}, timeout in {} seconds", order.getOrderUuid(), delay);
            set.add(new TimingTask(order, delay, cycleNum));
            cachedMap.put(order.getOrderUuid(), idx);
            log.info("idx: {}", idx);
            log.info("set: {}", set);
            log.info("cachedMap: {}", cachedMap);
        }
    }

    /**
     * 支付订单，将订单从环形队列中移除
     *
     * @param orderUuid 订单uuid
     *
     * @return
     */
    public boolean payOrder(String orderUuid){
        if (cachedMap.containsKey(orderUuid)) {
            Integer idx = cachedMap.get(orderUuid);
            OrderSet set = table[idx];
            if (set != null) {
                for (TimingTask task : set) {
                    if (orderUuid.equals(task.getOrderUuid())) {
                        synchronized (this) {
                            task.payment();
                            set.remove(task);
                            cachedMap.remove(orderUuid);
                            log.info("pay an order: {}");
                        }
                        break;
                    }
                }
            }
        }
        log.error("订单号不存在");
        return false;
    }

    /**
     *
     * @param idx
     *
     * @return
     */
    private int index(int idx) {
        return idx % DEFAULT_SIZE;
    }

    /**
     * 计算订单在循环队列中存在的圈数
     * @param idx 索引
     *
     * @return
     */
    private int cycleNumber(int idx) {
        return idx / DEFAULT_SIZE;
    }

    /**
     * 计算指针下一个应该指向的索引
     * @return
     */
    private int nextPoint() {
        int p = this.point.get();
        p = index(p + 1);
        return p;
    }

    /**
     * 封装订单的Wrapper
     */
    @Data
    private class TimingTask {

        /** 订单uuid */
        private String orderUuid;

        /** 订单 */
        private Order order;

        /** 延时，多少秒后执行任务 */
        private int delay;

        /** 订单 */
        private int cycleNum;

        /**
         * 包装订单，让其可以保存在循环队列的Set集合中
         *
         * @param order
         * @param delay
         * @param cycleNum
         */
        public TimingTask(Order order, int delay, int cycleNum) {
            this.order = order;
            this.orderUuid = order.getOrderUuid();
            this.delay = delay;
            this.cycleNum = cycleNum;
        }

        /**
         * 设置订单超时
         */
        public void timeout() {
            this.order.setTimeout(true);
            this.order.setUpdateTime(new Date());
        }

        /**
         * 支付订单
         */
        public void payment() {
            Date now = new Date();
            this.order.setPayed(true);
            this.order.setPayTime(now);
            this.order.setUpdateTime(now);
        }
    }

    /**
     * 放入循环队列的对象，实质是一个Set集合
     */
    private class OrderSet extends HashSet<TimingTask> {

    }

    /**
     * 轮询循环队列的任务，将当前Set集合取出交由别的线程扫描，队列指针加一
     */
    private static class RoundRinTimerTask implements Runnable {

        private CircularQueue circularQueue;
        private ExecutorService executor;

        public RoundRinTimerTask(CircularQueue circularQueue) {
            this.circularQueue = circularQueue;
            this.executor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.SECONDS,
                    new ArrayBlockingQueue<Runnable>(10),
                    Executors.defaultThreadFactory(),
                    new ThreadPoolExecutor.AbortPolicy());
        }

        /**
         * 对循环队列进行轮询，将扫描到的OrderSet对象让线程池去处理
         */
        @Override
        public void run() {
            synchronized (circularQueue) {
                // 循环队列当前指针
                int curPoint = circularQueue.point.get();
                // 循环队列指针下一个应该指向的索引
                int nextPoint = circularQueue.nextPoint();
                log.info("cur point: {}, next point: {}", curPoint, nextPoint);
                OrderSet set = circularQueue.table[curPoint];
                executor.submit(new TimeoutTask(circularQueue, set));
                circularQueue.point.set(nextPoint);
            }
        }
    }

    /**
     * 扫描Set集合的任务
     */
    private static class TimeoutTask implements Runnable {

        private CircularQueue circularQueue;
        private OrderSet set;

        public TimeoutTask(CircularQueue circularQueue, OrderSet set) {
            this.circularQueue = circularQueue;
            this.set = set;
        }

        /**
         * 遍历环形队列中的Set集合，移除超时的订单，为超时的订单cycleNum减一
         */
        @Override
        public void run() {
            if (set == null) {
                return;
            }
            Iterator<TimingTask> iterator = set.iterator();
            for (; iterator.hasNext();) {
                TimingTask task = iterator.next();
                int cycleNum = task.getCycleNum();
                if (cycleNum == 0) {
                    synchronized (circularQueue) {
                        iterator.remove();
                        task.timeout();
                        circularQueue.cachedMap.remove(task.getOrderUuid());
                        log.info("timeout task uuid: {}", task.getOrderUuid());
                        log.info("order timeout: {}", task.getOrder());
                        log.info("set: {}", set);
                        log.info("cachedMap: {}", circularQueue.cachedMap);
                    }
                } else {
                    task.setCycleNum(cycleNum - 1);
                }
            }
        }
    }
}


src/main/java/com/tuyu/controller/OrderController.java
package com.tuyu.controller;

import com.tuyu.common.CircularQueue;
import com.tuyu.entity.Order;
import com.tuyu.util.StringUtils;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * @author tuyu
 * @date 2/19/19
 * Talk is cheap, show me the code.
 */
@Slf4j
@RestController
public class OrderController {

    @Autowired
    private CircularQueue circularQueue;

    /**
     * 默认10秒
     */
    private static final int DELAY = 10;

    /**
     * 模拟添加一个订单，可以指定多少秒超时，如果不指定，默认10秒就超时
     * <pre>
     *     客户端请求：
     *          postman eg:
     *              localhost:8081/order
     *              POST
     *              name:香烟
     *              price:1000
     *              delay:5
     *          curl eg:
     *              curl -X POST --header 'Content-Type: application/x-www-form-urlencoded' --header 'Accept: text/xml' \
     *              -d 'name=香烟&price=1000&delay=8' 'http://localhost:8081/order'
     *
     * </pre>
     *
     * @param order 订单
     * @param delay 超时时间
     *
     * @return
     */
    @PostMapping("/order")
    public Object add(Order order, Integer delay) {
        if (order != null) {
            order.setOrderUuid(StringUtils.uuid());
            circularQueue.submitOrder(order, delay == null ? DELAY : delay);
            return "订单提交成功";
        }
        return "参数错误";
    }
}
